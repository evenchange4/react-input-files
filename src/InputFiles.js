// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import { Portal } from 'react-portal';

class InputFiles extends React.Component<{
  children: React.Node,
  onChange: (files: Array<Object>, e: SyntheticEvent<HTMLInputElement>) => void,
  accept?: string,
  style?: Object,
}> {
  static propTypes = {
    children: PropTypes.node.isRequired,
    onChange: PropTypes.func, // (files: Array<Object>, e: SyntheticEvent<HTMLInputElement>) => void,
    accept: PropTypes.string,
    style: PropTypes.object,
  };
  static defaultProps = {
    accept: 'image/*',
  };
  onInputRef = (node: any): void => {
    this.input = node;
  };
  onClick = (e: SyntheticEvent<HTMLButtonElement>) => {
    e.preventDefault();
    if (this.input) this.input.click();
  };
  onChange = (e: Object) => {
    this.props.onChange(e.target.files, e);
  };
  input: ?React.ElementRef<'input'>;
  render() {
    const { children, accept, style } = this.props;
    const { onInputRef, onClick, onChange } = this;

    return (
      <React.Fragment>
        <div
          onClick={onClick}
          onKeyPress={onClick}
          role="button"
          tabIndex="0"
          style={{ display: 'inline-block', ...style }}
        >
          {children}
        </div>

        <Portal>
          <input
            ref={onInputRef}
            type="file"
            accept={accept}
            onChange={onChange}
            style={{ display: 'none' }}
          />
        </Portal>
      </React.Fragment>
    );
  }
}

export default InputFiles;
