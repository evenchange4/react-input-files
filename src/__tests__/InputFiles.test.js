// @flow
import * as React from 'react';
import { mount } from 'enzyme';
import InputFiles from '../index';

it('should handle onClick without error', () => {
  const wrapper = mount(
    <InputFiles onChange={() => {}}>
      <div>input file</div>
    </InputFiles>,
  );
  const clickable = wrapper.find('div').first();
  clickable.simulate('click');
});

it('should handle onChange', () => {
  const mockOnChange = jest.fn();
  const wrapper = mount(
    <InputFiles onChange={mockOnChange}>
      <div>input file</div>
    </InputFiles>,
  );
  const input = wrapper.find('input').first();
  input.simulate('change', { target: { files: [{ name: 1 }, { name: 2 }] } });
  expect(mockOnChange).toHaveBeenCalled();
  expect(mockOnChange).toMatchSnapshot();
});
