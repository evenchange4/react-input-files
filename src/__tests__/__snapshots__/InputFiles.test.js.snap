// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle onChange 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Object {
          "name": 1,
        },
        Object {
          "name": 2,
        },
      ],
    ],
  ],
}
`;
